packages/task/src/browser/task-service.ts

```diff
@@ -402,10 +432,146 @@ export class TaskService implements TaskConfigurationClient {
+                })
+            );
+            await Promise.all(promises);
+        }
+        const taskInfo = await this.runTask(task, option);
+        if (taskInfo) {
+            if (task.isBackground) {
+                return Promise.race([this.getBackgroundTaskEndMatch(taskInfo.taskId), this.getExitCode(taskInfo.taskId)]).then(async result => {
+                    if (typeof result === 'number' && result !== 0) {
+                        await this.throwTaskError('The task: ' + task.label + ' terminated with exit code ' + result + '.');
+                    }
+                    return taskInfo;
+                });
+            } else {
+                const exitCode = await this.getExitCode(taskInfo.taskId);
+                if (exitCode !== 0) {
+                    await this.throwTaskError('The task: ' + task.label + ' terminated with exit code ' + exitCode + '.');
+                }
+            }
+        }
+        return taskInfo;
+    }
+
+    async throwTaskError(errorMessage: string): Promise<void> {
+        throw new Error(errorMessage);
+    }
+
+    /**
+     * Creates a graph of dependencies tasks from the root task and verify there is no DAG (Directed Acyclic Graph).
+     * In case of detection of a circular dependency, an error is thrown with a message which describes the detected circular reference.
+     */
+    detectDirectedAcyclicGraph(task: TaskConfiguration, tasknode: TaskNode, tasks: TaskConfiguration[]): void {
+        if (task && task.dependsOn && task.dependsOn.length > 0) {
+            for (let i = 0; i < task.dependsOn.length; i++) {
+                // It may be a string or a TaskIdentifier object
+                const childTaskIdentifier = task.dependsOn[i];
+                const childTaskConfiguration = this.getDependentTask(childTaskIdentifier, tasks);
+
+                // If current task and child task are identical or if
+                // one of the child tasks is identical to one of the current task ancestors, then raise an error
+                if (this.taskDefinitionRegistry.compareTasks(task, childTaskConfiguration) ||
+                    tasknode.parentsID.filter(t => this.taskDefinitionRegistry.compareTasks(childTaskConfiguration, t)).length > 0) {
+                    const fromNode = task.label;
+                    const toNode = childTaskConfiguration.label;
+                    throw new Error('Circular reference detected: ' + fromNode + ' -->  ' + toNode);
+                }
+                const childnode = new TaskNode(childTaskConfiguration, [], Object.assign([], tasknode.parentsID));
+                childnode.addParentDependency(tasknode.taskId);
+                tasknode.addChildDependency(childnode);
+                this.detectDirectedAcyclicGraph(childTaskConfiguration, childnode, tasks);
+            }
+        }
+    }
+
+    // tslint:disable-next-line: no-any
```

Comment:
parameter and return types should not use `any`, only if it json, please review the rest PR to have explicit types

Reply:
It might be a json object.
was fixed by renaming the variable and a documentation was added to the method